### ISSUE TITLE
MacOS ERROR: The Python ssl extension was not compiled. Missing the OpenSSL lib?

### ISSUE THREAD

**Issue by qushengswaych** [pyenv/pyenv#1566]
When I want to install python 2.7.10 on my Mac, I got an error.
How to do with this?
Command:
`CFLAGS="-I$(brew --prefix openssl)/include" LDFLAGS="-L$(brew --prefix openssl)/lib" pyenv install 2.7.10`
Result:
```
WARNING: Please make sure you remove any previous custom paths from your /Users/qusheng/.pydistutils.cfg file.
python-build: use openssl from homebrew
python-build: use readline from homebrew
Downloading Python-2.7.10.tar.xz...
-> https://www.python.org/ftp/python/2.7.10/Python-2.7.10.tar.xz
Installing Python-2.7.10...
patching file ./Lib/site.py
python-build: use readline from homebrew
python-build: use zlib from xcode sdk
ERROR: The Python ssl extension was not compiled. Missing the OpenSSL lib?

Please consult to the Wiki page to fix the problem.
https://github.com/pyenv/pyenv/wiki/Common-build-problems


BUILD FAILED (OS X 10.15.3 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/7k/88slwqdn6w3_3bq688fzswtw0000gn/T/python-build.20200323153019.44397
Results logged to /var/folders/7k/88slwqdn6w3_3bq688fzswtw0000gn/T/python-build.20200323153019.44397.log

Last 10 log lines:
rm -f /Users/qusheng/.pyenv/versions/2.7.10/share/man/man1/python.1
(cd /Users/qusheng/.pyenv/versions/2.7.10/share/man/man1; ln -s python2.1 python.1)
if test "xno" != "xno"  ; then \
		case no in \
			upgrade) ensurepip="--upgrade" ;; \
			install|*) ensurepip="" ;; \
		esac; \
		 ./python.exe -E -m ensurepip \
			$ensurepip --root=/ ; \
	fi
```

Platform information:macOS Catalina 10.15.3 (19D76)
pyenv version:1.2.17
Python version:2.7.17

**Comment by joshfriend** [comment]
please check the common build problems wiki mentioned in your error logs, and also provide the info requested by the issue template

**Comment by doctorduke** [comment]
Okay. I had a different experience. I'm just cross posting this in the locations that partially helped me so if someone ends up there related to this it might help. Sorry if this isn't helpful to this direct train of thought or what helps you specifically.
This was all wrote in Notion so I apologize if it's formatting is off. Here's the link to it in my Notion. You can send me comments there and I should get them quickly. Hit me up however you can find a way if the link gets broken. I'll more than likely update there before I update everywhere else.

https://www.notion.so/Working-on-getting-python-and-pip-821f4b2d90034f7595ef5470f6b80286

Pre-preface: pardon my dark humour but I have had a few days of banging my head against a wall on a few other things and this was the icing on the cake.


To preface what I have been trying to do is install basic command line tools for the languages I use and are called for in some new things I'm trying to learn on a fresh OSX install. I'm doing this on two fresh Big Sur installs because I like to keep my machines in parity. The idea of that is to be able to easily swap between my machines if one is running something intense and I want to move on to something else.

This is done on Big Sur 11.0.1 and brew version 2.7.7.

I'm trying to give as comprehensive a version of the final steps and not the hours upon hours I have spent working to fix this.

Also, I'd like to mention I'm familiar with these tools in no way an expert so this was not fun and I imagine any new user being given a new computer and having to do this would be so freaking discouraged its gross. Beyond that I'm on a 4Mbit connection because I have moved to a rural area with no good alternatives. So this has been excruciating having to also deal with failed huge downloads after hours of waiting for them to finish. Having previously mentioned that I am doing this on 2 machines I'd like to also point out how painful a system like the Mac App Store is when you have to spend hours just to install Xcode and then do it again on a second machine because 1. There doesn't seem to be a straighforward way to share locally your installs 2. I'm sure there are other ways to do direct downloads and other tricks but I wasn't trying to solve that problem. I was just trying to install 2 versions of python and tools related to it.

In the end using 'pyenv' I was able to install Python 3.8.0. That let me see that there was some hope but I was trying to install a Python 2.7.x version. I was able to get both installed but need to try again from scratch to see if they can work in a different order.

I have had many many errors come through. I may not get them all but I had this warning

So to be honest I can't even remember what the task I was setting out to do was. OH. I was trying to get started working with [hass.io](http://hass.io) (Home Assistant) and install it on one of my Macs. I was told to use the command 'virtual-env'. I can't remember the specifics because I've been at this 8 hours today and tried what feels like a million things. If you could see how great a noose I worked on while waiting for failures! (I kid.) Somehow virtual-env lead me to needing to install pip I think. I can't remember if it was trying to get pip or pyenv installed but there were a bunch of errors about permissions. I must have cleaned my log but they were errors about permissions for /usr/local.

## Permission errors installing pyenv

The first step to doing this I found was disabling SIP (System Integrity Protection). I was hoping to find a way to do this from the command line but all the examples of how to do this didn't work for me. So I ended up 

1. Shut down your Mac
2. Ensure your Mac has shut down.
3. It probably didn't.
4. Do it again.
5. Nope.
6. Just hold the power button for 4 minutes and hope it works.
7. Once it's shutdown Hold CMD+R 
8. Then in addition to having CMD+R held press the power button until it seems like your Mac may finally start to boot. Its different every time!
9. You can and maybe should release CMD+R when you see the Apple logo. If it worked you should see a large apple if running at a resolution during normal operations.
10. Select your language. (I guess now is the best time to mention barring anything blowing my mind this will all be in English.)
11. You'll get presented with a menu eventually that has a bunch of options one of which will be "Disk Utility". You're almost through Step -10!
12. At the top menu under Utilities select terminal.
13. In the terminal window that I hope for your has opened input the following command.

    ```bash
    csrutil disable
    ```

    - credit stackoverflow post

        [https://stackoverflow.com/questions/63312303/is-there-a-way-to-change-a-read-only-file-system-to-a-writable-file-system-on-ma](https://stackoverflow.com/questions/63312303/is-there-a-way-to-change-a-read-only-file-system-to-a-writable-file-system-on-ma)

14. Then you can restart. I'll note everything I ever read says turn it back on after. Hell no in my opinion. This setting has never once to my knowledge done anything but cause me headaches and waste hours upon hours of my time. {EXPLETIVE DELETED} that.

Now that SIP (more like BLOWS) is disabled you can proceed to your next error.

### You still can't modify the permission on this directory

Now you are able to modify the read write permissions on the contents of the /usr/local directory. This is accomplished with the following.

```bash
sudo chmod -R 775 /usr/local
```

You'd think that would be enough but 

```bash
Error: The following directories are not writable by your user:
/usr/local/lib/pkgconfig
/usr/local/share/info
/usr/local/share/man/man3
/usr/local/share/man/man5
```

Luckily the output about this gives you the answer, at least in my situation it did.

1. Enter the following in your terminal 

```bash
sudo chown -R $(whoami) /usr/local/lib/pkgconfig /usr/local/share/info /usr/local/share/man/man3 /usr/local/share/man/man5
```

Congrats! You have only begun to understand the frustration if you were trying to get further than this and ended up here.

You can now install pyenv with brew.

```bash
brew install pyenv
```

Don't know how to help you if that didn't work. 

## What ended up being Python 3.8.0 instead of 2.7.9 installation steps and errors.

So from here. Again let me say I'm not an expert in this stuff. I have a bad habit of not reading every line of spam that comes out when working in the command line and the process takes forever. From here I did follow what it said about some errors with OpenSSL and added those to my bash_profile and sourced it in the window. Something I'll say may not be glaringly obvious to some that aren't pros here is that the output specifically says this.

```bash
If you need to have openssl@1.1 first in your PATH, run:
  echo 'export PATH="/usr/local/opt/openssl@1.1/bin:$PATH"' >> /Users/doctorduke/.bash_profile

For compilers to find openssl@1.1 you may need to set:
  export LDFLAGS="-L/usr/local/opt/openssl@1.1/lib"
  export CPPFLAGS="-I/usr/local/opt/openssl@1.1/include"

For compilers to find readline you may need to set:
  export LDFLAGS="-L/usr/local/opt/readline/lib"
  export CPPFLAGS="-I/usr/local/opt/readline/include"

For pkg-config to find readline you may need to set:
  export PKG_CONFIG_PATH="/usr/local/opt/readline/lib/pkgconfig"
```

To some you may be like "Obviously." To other who don't do this all the time and/or it's rare you have to you might get caught up doing exactly what this thing says and setting. NOT appending these values when both are needed. IT SPECIFICALLY says set =. That's an assignment to an exact value.  When everything is failing you probably try and follow the instructions to a T.

So I add a few experimental things to my bash_profile and without updating it to what was in the end that worked it had these in it related to getting pyenv to install (what at this point was still an attempt at 2.7.9).

```bash
eval "$(pyenv init -)"
eval "$(pyenv virtualenv-init -)"
export LDFLAGS="-L/usr/local/opt/zlib/lib -L/usr/local/opt/bzip2/lib"
export CPPFLAGS="-I/usr/local/opt/zlib/include -I/usr/local/opt/bzip2/include"

export PATH="$HOME/.pyenv/bin:$PATH"
export PATH="/usr/local/bin:$PATH"

export PATH="/usr/local/opt/sqlite/bin:$PATH"
export PKG_CONFIG_PATH="/usr/local/opt/sqlite/lib/pkgconfig
```

Again, this has been hours of working with extremely similar modifications across multiple machines. I'm trying to hit every error I saw and what got around it but I doubt I will. Some parts may be out of order or missed because I cleaned some consoles a couple times and overwrote a few things when something failed too many times.

### LOTS of messages about zlib not working.

So if you go the route of doing the xcode command line tools. On Big Sur 11.0.1 I ended up, because one of the many results that said it solved my problem (which this did not seem to directly solve my problem) but I encountered a bunch of steps that there are bugs in versions of the command line tools around version 12.2 in beta or and a few others. I ended up wasting 2 hours of my life having to upgrade to Xcode 12.4 (separately on 2 machines) and the command line tool (this went okay because I found a direct link to the .dmg and was able to Airdrop it to my second machine when I verified the download. [https://download.developer.apple.com/Developer_Tools/Command_Line_Tools_for_Xcode_12.4/Command_Line_Tools_for_Xcode_12.4.dmg](https://download.developer.apple.com/Developer_Tools/Command_Line_Tools_for_Xcode_12.4/Command_Line_Tools_for_Xcode_12.4.dmg)

Further steps say two options I believe are the same thing but I'm not sure. One is use xcode-select and blah. The following was the next step in getting me somewhere.

Via this link

[https://github.com/pyenv/pyenv/issues/1737](https://github.com/pyenv/pyenv/issues/1737)

(Oh yeah. PS. I hit up the freaking Common Build Errors page. Thanks for nothing the guys that go out of their way to say this when most people have already done it. You don't need to tell people to do this over and over when it's right there in the output. The info isn't always up to date and again FRESH INSTALL of everything so it should be straight forward. Still none of this info was compiled there.)

What did help was this awesome dude at the above link. Per him in that link the following two steps got me (and these are exactly following his instructions. Which is why I got Python 3.8.0)

## **Step 1 Align command-line tools**

Open Xcode-beta.appGo to Preference > LocationsSelect the right version of command-line tools

## **Step 2 Install python**

In this case, I installed 3.8.0. If you want to install a different version, you will need to change the version in the following command.`CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" pyenv install --patch 3.8.0 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)`

I just copied and pasted to not screw anything up.

But this seemed to work! The wrong version, some version of 3 which wasn't what I was trying to get to yet. Now I'm ahead of the game though. Now let's just change the version number as mentioned further in the post to the one we were going for and it'll be good.

## Crap I forgot about this and only remembered when I was almost at the end

I execute this command before I got 3.8.0 to install properly. Not sure what layer of work it helped but I vaguely remember it getting rid of something. 

```bash
brew reinstall zlib bzip2
```

## Python 2.7.9 steps and errors.

```bash
CFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include" LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib" pyenv install --patch 2.7.9 < <(curl -sSL https://github.com/python/cpython/commit/8ea6353.patch\?full_index\=1)
```

Yay right? No. Not yay. More of the same about the OpenSSL being wrong. 

From here I'm going to spare you an even more long winded explanation about all the different CFLAGS and LFLAGS I tried to vary between to get this to work. 

At some point I was noticing before the OpenSSL error that there was a warning about SQlite3 not compiling so I decided to fix that incase it was causing me issues with OpenSSL. This didn't help the exact error I was getting but getting rid of warnings makes the errors easier to distinguish. 

```bash
brew install sqlite3
```

### After getting SQLite3 reinstalled correctly.

```bash
ERROR: The Python ssl extension was not compiled. Missing the OpenSSL lib?

Please consult to the Wiki page to fix the problem.
https://github.com/pyenv/pyenv/wiki/Common-build-problems

BUILD FAILED (OS X 11.0.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/6z/cngcgtk97zvdh818_gg5qytc0000gn/T/python-build.20210204152125.11840
Results logged to /var/folders/6z/cngcgtk97zvdh818_gg5qytc0000gn/T/python-build.20210204152125.11840.log
```

Okay. The above may not be the final error. There's a few variations of the ones I was getting. This one was unique.

```bash
Py_FatalError("abort() called from Python code didn't abort!");
    ^~~~~~~~~~~~~
1 warning and 1 error generated.
make: *** [Modules/posixmodule.o] Error 1
make: *** Waiting for unfinished jobs....

BUILD FAILED (OS X 11.0.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/6z/cngcgtk97zvdh818_gg5qytc0000gn/T/python-build.20210204134530.6409
Results logged to /var/folders/6z/cngcgtk97zvdh818_gg5qytc0000gn/T/python-build.20210204134530.6409.log
```

I think that might have been trying to do a 2.6 version though.

Here is another

```bash
ERROR: The Python ssl extension was not compiled. Missing the OpenSSL lib?

Please consult to the Wiki page to fix the problem.
https://github.com/pyenv/pyenv/wiki/Common-build-problems

BUILD FAILED (OS X 11.0.1 using python-build 20180424)

Inspect or clean up the working tree at /var/folders/6z/cngcgtk97zvdh818_gg5qytc0000gn/T/python-build.20210204151421.87649
Results logged to /var/folders/6z/cngcgtk97zvdh818_gg5qytc0000gn/T/python-build.20210204151421.87649.log

Last 10 log lines:
rm -f /Users/doctorduke/.pyenv/versions/2.7.9/share/man/man1/python.1
(cd /Users/doctorduke/.pyenv/versions/2.7.9/share/man/man1; ln -s python2.1 python.1)
if test "xno" != "xno"  ; then \
		case no in \
			upgrade) ensurepip="--upgrade" ;; \
			install|*) ensurepip="" ;; \
		esac; \
		 ./python.exe -E -m ensurepip \
			$ensurepip --root=/ ; \
	fi
```

Yet another version

```bash
Installing Python-2.7.7...

ERROR: The Python zlib extension was not compiled. Missing the zlib?

Please consult to the Wiki page to fix the problem.
https://github.com/pyenv/pyenv/wiki/Common-build-problems

BUILD FAILED
```

Eventually I found this post that gave me some info further. I read through it and a few other pages and it seemed to me that the OpenSSL from brew was not working with this version of python.

[https://github.com/pyenv/pyenv/issues/1669](https://github.com/pyenv/pyenv/issues/1669)

This post references this command

```bash
PYTHON_BUILD_HOMEBREW_OPENSSL_FORMULA=openssl@1.0 pyenv install 2.6.9
```

This appeared to work for the guy. Alas it did not work for me on that version or 2.7.9. So I screwed around with a few commands a bit more and nothing seemed to work. EVENTUALLY, I aged 400 years and I had some nicotine I was able to clear a bit of how clouded my head was from anger and frustration and I looked at the output from that command and the complaints about OpenSSL from the line that worked on 3.8.0. So what I ended up doing was stripping some commands from the previously mentioned command to install 3.8.0. I got rid of all info surrounds ssl in that command and ended up just defining my CFLAGS and LFLAGS as such

```bash
CFLAGS="-I$(brew --prefix bzip2)/include -I$(brew --prefix readline)/include -I$(xcrun --show-sdk-path)/usr/include -I/usr/local/opt/sqlite/include" 
LDFLAGS="-L$(brew --prefix readline)/lib -L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib -L/usr/local/opt/sqlite/lib"
```

Notice the stuff around ssl gone. Now with these flags set this way executed

```bash
PYTHON_BUILD_HOMEBREW_OPENSSL_FORMULA=openssl@1.0 pyenv install 2.7.9
```

The following I was so exhausted from rage I couldn't even celebrate. My whole world was shaken because I had no great battles left to fight. My purpose was over.

```bash
mac256g:~ doctorduke$ PYTHON_BUILD_HOMEBREW_OPENSSL_FORMULA=openssl@1.0 pyenv install 2.7.9
Error: No available formula with the name "openssl@1.0".
Downloading openssl-1.0.2k.tar.gz...
-> https://pyenv.github.io/pythons/6b3977c61f2aedf0f96367dcfb5c6e578cf37e7b8d913b4ecb6643c3cb88d8c0
Installing openssl-1.0.2k...
Installed openssl-1.0.2k to /Users/doctorduke/.pyenv/versions/2.7.9

python-build: use readline from homebrew
Downloading Python-2.7.9.tar.xz...
-> https://www.python.org/ftp/python/2.7.9/Python-2.7.9.tar.xz
Installing Python-2.7.9...
patching file ./Lib/site.py
patching file ./Lib/ssl.py
python-build: use readline from homebrew
python-build: use zlib from xcode sdk
Installed Python-2.7.9 to /Users/doctorduke/.pyenv/versions/2.7.9
```

I now have no idea what to do. I guess I could catch up on the classes I had scheduled for the day but had to skip. Maybe I could take another step toward playing with Home Assistant. If you can call over 8 hours of wanting to drive a knife through your temple "playing". At this point I'm afraid of snakes and worried if I even try to run some python it will give me an error and I'll have to throw myself off the roof.

Any way. I hope you enjoyed my dark humour around a frustrating issue and my apparent incompetence. If I know engineers on the internet I'll hear some less than polite words about my stupidity either directly or indirectly. But I just really hope I can spare someone the frustration and stolen time so they can take their classes.

### Few things at the end I can't remember the details of.

I'm just going to throw this at the end but this was what I was originally trying to do and at some point something complained about this and I followed the steps on this page [https://github.com/pyenv/pyenv-virtualenv/issues/25](https://github.com/pyenv/pyenv-virtualenv/issues/25) to get around it. Blanking on where that happened. If you encounter it and I can update this post let me know.

```bash
brew install --HEAD pyenv-virtualenv
```

Let me just keep reminding you, when everything is inexplicably failing sometimes you have to go slightly braindead and just follow the instructions to find your way back. So I know some of these steps are "Duh." "Idiot doesn't know." "Amateur." But EVERY STEP WAS A PAIN and even the resolution steps were unclear or only worked in some cases. 

### Close

If there are edits to this process let me know. I'm not one that normally posts things out there for people to see because typically you're just inviting a chance to get insulted by someone smarter than you at working with these things but not smart enough to realize some people just don't encounter these situations and some people make mistakes and overlook things. Most of the time this stuff is put out the current fire and deal with the next fire when it occurs and I'm currently in a place where I'm trying to get all the fires out of the way so I can be better informed so when the next fire happens it's not quite as big. If you experience some variations because of a changed Mac OS version or different pyenv or Homebrew or if you just want to say hi, let me know. I'll try and update this post whereever I end up putting it or I'll say hi back. If you come across better versions of the errors or more informative ones than I included I'll do that too. If you have the inkling to help make it easier for others down the road. Also please keep in mind everyone starts somewhere and again, some kid could be playing with his first computer he just got and wants to play with Python to learn something having never used a single tool and he could encounter this. So many steps out there tell you what one versions solution was but a that kid's setup may not work with that version. I don't know, that's just how I look at it.

Oh man! I almost forgot. If you know how to fix Chrome always launching new windows in the top left when you want it in the bottom right corner GOD please let me know. It's killing me. My monitor is a 55 inch tv so that I can treat it like 4 screens to have reference stuff up and it drives me nuts!!

**Comment by qeleb** [comment]
Make sure to use `openssl@1.1` rather than putting `openssl` or `openssl@3` if you are installing Python <3 (like 2.7.18)

If you are using brew, add this to `.zshrc`/`.bashrc` for installing Python <3:
```
export LDFLAGS="-L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib -L$(brew --prefix openssl@1.1)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix ncurses)/lib -L$(xcrun --show-sdk-path)/usr/lib"
export CFLAGS="-I$(brew --prefix zlib)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix openssl@1.1)/include -I$(brew --prefix readline)/include -I$(brew --prefix ncurses)/include -I$(xcrun --show-sdk-path)/usr/include"
export CPPFLAGS="-I$(brew --prefix zlib)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix openssl@1.1)/include -I$(brew --prefix readline)/include -I$(brew --prefix ncurses)/include -I$(xcrun --show-sdk-path)/usr/include"
```

If you are using brew, add this to `.zshrc`/`.bashrc` for installing Python >=3:
```
export LDFLAGS="-L$(brew --prefix zlib)/lib -L$(brew --prefix bzip2)/lib -L$(brew --prefix openssl)/lib -L$(brew --prefix readline)/lib -L$(brew --prefix ncurses)/lib -L$(xcrun --show-sdk-path)/usr/lib"
export CFLAGS="-I$(brew --prefix zlib)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix openssl)/include -I$(brew --prefix readline)/include -I$(brew --prefix ncurses)/include -I$(xcrun --show-sdk-path)/usr/include"
export CPPFLAGS="-I$(brew --prefix zlib)/include -I$(brew --prefix bzip2)/include -I$(brew --prefix openssl)/include -I$(brew --prefix readline)/include -I$(brew --prefix ncurses)/include -I$(xcrun --show-sdk-path)/usr/include"
```
